onadata/libs/serializers/metadata_serializer.py

```diff
@@ -51,14 +60,27 @@ def create(self, validated_data):
         data_type = validated_data.get('data_type')
         data_file = validated_data.get('data_file')
         xform = validated_data.get('xform')
-        data_value = data_file.name if data_file else validated_data.get('data_value')
-        data_file_type = data_file.content_type if data_file else None
+        data_value = (
+            data_file.name if data_file else validated_data.get('data_value')
+        )
+
+        allowed_types = settings.SUPPORTED_MEDIA_UPLOAD_TYPES
+        content_type = (
+            data_file.content_type
+            if data_file and data_file.content_type in allowed_types
+            else mimetypes.guess_type(data_value)[0]
+        )
+
+        if content_type not in allowed_types:
+            raise serializers.ValidationError(
+                {'content_type': _('Invalid content type.')}
+            )
```

Comment:
Because we are inside the serializer,  I would have preferred to use a DRF validating methods instead, e.g.: `def validate_data_file_type(self, value)`.  
The validation could take place inside the `validate()` method.

Reply:
@joshuaberetta made me notice that magical method `validate_data_file_type` would not be trigger because `data_file_type` is not a required field and it is not part of `Meta.fields`. 
We can leave it as it for the moment.

In KoBoCAT 2.0, this changes. We can open an issue to implement this validation.